### PR TITLE
[wip]CRM-21224 reinstate use of limit on dedupe searches + other fixes

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Adding limit=50 to a dedupe url will restrict the search to only finding matches for 50 contacts. This can be important on a large site but is a largely hidden feature that broke recently in a dedupe tidy up.

Before
----------------------------------------
issue 1) Deduping with limit=50 in the url will dedupe the whole db. 
issue 2) After doing a batch merge all the conflicted rows do not  show up on the conflicts screen

After
----------------------------------------
1) Deduping with limit=50 in the url will only find matches for the first 50 contacts 
2) the conflicts screen shows a list of merges that hit conflicts.

Technical Details
----------------------------------------
There is some tidy up in here - I will comment on the lines in detail but basically the goal is to rationalise some parameters, notably the bad way $batchLimit is handled as a magic param combined with it being added to the $where clause. I have tidied that up & I'm fairly sure that I have maintained it doing what it was doing .... and that it is & was slightly broken ....

---

 * [CRM-21224: Dedupe searches dropping limit](https://issues.civicrm.org/jira/browse/CRM-21224)

### CODE REVIEW

File: CRM/Dedupe/Finder.php [hunk 1]
```diff
@@ -51,28 +51,29 @@ class CRM_Dedupe_Finder {
...
+   *  the number of searched contacts, not the matches found.
    *
    * @return array
    *   Array of (cid1, cid2, weight) dupe triples
    *
    * @throws CiviCRM_API3_Exception
    * @throws Exception
    */
-  public static function dupes($rgid, $cids = array(), $checkPermissions = TRUE, $limit = NULL) {
+  public static function dupes($rgid, $cids = array(), $checkPermissions = TRUE, $searchLimit = 0) {
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Finder.php]: changing to 0 is more consistent with elsewhere

File: CRM/Dedupe/Merger.php [hunk 2]
```diff
@@ -619,19 +619,19 @@ public static function retrieveFields($main, $other) {
    *
    * @return array|bool
    */
-  public static function batchMerge($rgid, $gid = NULL, $mode = 'safe', $batchLimit = 1, $isSelected = 2, $criteria = array(), $checkPermissions = TRUE) {
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: batchMerge is never actually called without $batchLimit being passed so we can change or alter the default. Coin toss as to which but I came down on changing to 0 which means no limit

File: CRM/Dedupe/Merger.php [hunk 3]
```diff
@@ -619,19 +619,19 @@ public static function retrieveFields($main, $other) {
...
-    $redirectForPerformance = ($batchLimit > 1) ? TRUE : FALSE;
-    $reloadCacheIfEmpty = (!$redirectForPerformance && $isSelected == 2);
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: what is this dark magic?

File: CRM/Dedupe/Merger.php [hunk 4]
```diff
@@ -619,19 +619,19 @@ public static function retrieveFields($main, $other) {
...
+    $reloadCacheIfEmpty = ($batchLimit <= 1 && $isSelected == 2);
     $dupePairs = self::getDuplicatePairs($rgid, $gid, $reloadCacheIfEmpty, $batchLimit, $isSelected, '', ($mode == 'aggressive'), $criteria, $checkPermissions);
 
     $cacheParams = array(
       'cache_key_string' => self::getMergeCacheKeyString($rgid, $gid, $criteria, $checkPermissions),
       // @todo stop passing these parameters in & instead calculate them in the merge function based
       // on the 'real' params like $isRespectExclusions $batchLimit and $isSelected.
       'join' => self::getJoinOnDedupeTable(),
-      'where' => self::getWhereString($batchLimit, $isSelected),
+      'where' => self::getWhereString($isSelected),
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: This is one of 2 places getWhereString is called from - pass batchLimit to the merge function & on to the retrieval function instead...

File: CRM/Dedupe/Merger.php [hunk 5]
```diff
@@ -619,19 +619,19 @@ public static function retrieveFields($main, $other) {
...
     );
-    return CRM_Dedupe_Merger::merge($dupePairs, $cacheParams, $mode, $redirectForPerformance, $checkPermissions);
+    return CRM_Dedupe_Merger::merge($dupePairs, $cacheParams, $mode, $batchLimit, $checkPermissions);
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: This is the only place this function is called with the last 2 parameters. The api calls it but with only the first 3

File: CRM/Dedupe/Merger.php [hunk 6]
```diff
@@ -652,21 +652,15 @@ public static function getJoinOnDedupeTable() {
...
    */
-  protected static function getWhereString($batchLimit, $isSelected) {
+  protected static function getWhereString($isSelected) {
     $where = "de.id IS NULL";
     if ($isSelected === 0 || $isSelected === 1) {
       $where .= " AND pn.is_selected = {$isSelected}";
     }
-    // else consider all dupe pairs
-    // @todo Adding limit to Where??!!
-    if ($batchLimit) {
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: Out damned spot

File: CRM/Dedupe/Merger.php [hunk 7]
```diff
@@ -768,16 +762,16 @@ public static function getMergeStatsMsg($cacheKeyString) {
...
+   *   Limit number of matches to process.
    *
    * @param bool $checkPermissions
    *   Respect logged in user permissions.
    *
    * @return array|bool
    */
   public static function merge($dupePairs = array(), $cacheParams = array(), $mode = 'safe',
-     $redirectForPerformance = FALSE, $checkPermissions = TRUE
+     $batchLimit = 0, $checkPermissions = TRUE
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: $redirectForPerformance is only passed in from one place

Comment (eileenmcnaughton) [reply CRM/Dedupe/Merger.php]: Note to self - test UI batch mode for this

**Comment by eileenmcnaughton** [comment]
@JKingsnorth Looks like limit got dropped in the last tidy up - but it's back. I have tried to tackle the weirdness & confusingness about limit handling a little too - although I think I need coffee + a re-read of the redirectForPerformance / batch = 1 vs 2 stuff. I think it might have been already broken...

@ejegg ....

File: CRM/Dedupe/Merger.php [hunk 8]
```diff
@@ -813,14 +811,14 @@ public static function merge($dupePairs = array(), $cacheParams = array(), $mode
...
-      if ($cacheKeyString && !$redirectForPerformance) {
+      if ($cacheKeyString && $isProcessUntilAllProcessed) {
         // retrieve next pair of dupes
         // @todo call getDuplicatePairs.
         $dupePairs = CRM_Core_BAO_PrevNextCache::retrieve($cacheKeyString,
           $cacheParams['join'],
           $cacheParams['where'],
           0,
-          0,
+          $batchLimit,
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: instead of hacking into 'where' pass this here

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@jmcclelland  @JKingsnorth any chance of one of you taking a look at this?

**Comment by eileenmcnaughton** [comment]
@JKingsnorth I've added another fix into this & fixed the tests - any chance you can take a look?

**Comment by JKingsnorth** [comment]
Hi @eileenmcnaughton sorry, this is on my radar but currently 4th on my list! Need more hours in the day.

**Comment by jmcclelland** [comment]
Thanks Eileen for sorting this all out. I tried to review the code as carefully as I could - I didn't see anything that stands out as off, but I had a hard time parsing it all. I also tried to back port it to the version PowerBase is on (4.7.19) so I could test it on one of our data sets, but too many conflicts with changes made since then.

**Comment by eileenmcnaughton** [comment]
@jmcclelland I hear you - I've put this to WIP & pulled out most of the changes (which started mounting up!) to a bunch of separate PRs. Once they have gone through I'll worry about the refactor in here & handling of $criteria params

**Comment by eileenmcnaughton** [comment]
I'm closing this - I've added some extra commits to the branch and it is actually now kinda old & I've moved my focus off this

**Pull request closed without merging**